indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/supervisor/autoscaler/CostBasedAutoScaler.java

```diff
@@ -0,0 +1,287 @@
+
+    if (validTaskCounts.length == 0) {
+      log.warn("No valid task counts after applying constraints for dataSource [%s]", supervisorId);
+      return -1;
+    }
+
+    // If idle is already in the ideal range [0.2, 0.6], optimal utilization has been achieved.
+    // No scaling is needed - maintain stability by staying at current task count.
+    final double currentIdleRatio = metrics.getPollIdleRatio();
+    if (currentIdleRatio >= 0 && WeightedCostFunction.isIdleInIdealRange(currentIdleRatio)) {
+      log.info(
+          "Idle ratio [%.3f] is in ideal range for dataSource [%s], no scaling needed",
+          currentIdleRatio,
+          supervisorId
+      );
+      return -1;
+    }
+
+    // Update bounds with observed lag BEFORE optimization loop
+    // This ensures normalization uses historical observed values, not predicted values
+    costFunction.updateLagBounds(metrics.getAvgPartitionLag());
+
+    int optimalTaskCount = -1;
+    double optimalCost = Double.POSITIVE_INFINITY;
+
+    final int bestTaskCountIndex = Arrays.binarySearch(validTaskCounts, currentTaskCount);
+    for (int i = bestTaskCountIndex - SCALE_FACTOR_DISCRETE_DISTANCE;
+         i <= bestTaskCountIndex + SCALE_FACTOR_DISCRETE_DISTANCE; i++) {
+      // Range check.
+      if (i < 0 || i >= validTaskCounts.length) {
+        continue;
+      }
+      int taskCount = validTaskCounts[i];
+      if (taskCount < config.getTaskCountMin()) {
+        continue;
+      } else if (taskCount > config.getTaskCountMax()) {
+        break;
+      }
+      double cost = costFunction.computeCost(metrics, taskCount, config);
+      log.debug("Proposed task count: %d, Cost: %.4f", taskCount, cost);
+      if (cost < optimalCost) {
+        optimalTaskCount = taskCount;
+        optimalCost = cost;
+      }
+    }
+
+    emitter.emit(metricBuilder.setMetric(OPTIMAL_TASK_COUNT_METRIC, (long) optimalTaskCount));
+
+    log.info(
+        "Cost-based scaling evaluation for dataSource [%s]: current=%d, optimal=%d, cost=%.4f, "
+        + "avgPartitionLag=%.2f, pollIdleRatio=%.3f",
+        supervisorId,
+        metrics.getCurrentTaskCount(),
+        optimalTaskCount,
+        optimalCost,
+        metrics.getAvgPartitionLag(),
+        metrics.getPollIdleRatio()
+    );
+
+    if (optimalTaskCount > currentTaskCount) {
+      return optimalTaskCount;
+    } else if (optimalTaskCount < currentTaskCount) {
+      supervisor.getIoConfig().setTaskCount(optimalTaskCount);
+    }
+    return -1;
+  }
+
+  /**
+   * Generates valid task counts based on partitions-per-task ratios.
+   * This enables gradual scaling and avoids large jumps.
+   *
+   * @return sorted list of valid task counts within bounds
+   */
+  static int[] computeFactors(int partitionCount)
+  {
+    if (partitionCount <= 0) {
+      return new int[]{};
+    }
+
+    List<Integer> result = new ArrayList<>();
+
+    for (int partitionsPerTask = partitionCount; partitionsPerTask >= 1; partitionsPerTask--) {
```

Comment:
We shouldn't be computing all possible valid task counts if we are only ever going to use the task counts in the +2/-2 window.

Please try to simplify this computation to something similar to the one suggested in https://github.com/apache/druid/pull/18819#discussion_r2605256520.